src/store/MinHeapLib.sol

```diff
@@ -0,0 +1,196 @@
+    ) internal view returns (uint256) {
+        return heap.heapIndexes[proofRequestIdentifier.chainId][proofRequestIdentifier.blockNumber];
+    }
+
+    // ---------------------------- Mutations ----------------------------
+
+    function addProofRequest(
+        Heap storage heap,
+        uint256 key,
+        IProofManager.ProofRequestIdentifier memory proofRequestIdentifier
+    ) internal {
+        // Ensure sentinel at index 0
+        if (heap.nodes.length == 0) {
+            heap.nodes.push(); // empty slot at index 0
+        }
+
+        heap.nodes.push(Node({ key: key, proofRequestIdentifier: proofRequestIdentifier }));
+        uint256 idx = heap.nodes.length - 1; // 1-based
+
+        heap.heapIndexes[proofRequestIdentifier.chainId][proofRequestIdentifier.blockNumber] = idx;
+
+        _siftUp(heap, idx);
+    }
+
+    function extractMin(Heap storage heap) internal returns (Node memory minNode) {
+        require(size(heap) > 0, "Heap: empty");
+
+        minNode = heap.nodes[1];
+
+        // Only one real element (sentinel + root)
+        if (heap.nodes.length == 2) {
+            heap.nodes.pop(); // remove root, sentinel remains
+        } else {
+            // Move last to root
+            Node memory moved = heap.nodes[heap.nodes.length - 1];
+            heap.nodes[1] = moved;
+            heap.nodes.pop();
+
+            // Update mapping for moved node
+            heap.heapIndexes[
+                moved.proofRequestIdentifier.chainId
+            ][moved.proofRequestIdentifier.blockNumber] = 1;
+
+            _siftDown(heap, 1);
+        }
+
+        // Clear mapping for removed node
+        heap.heapIndexes[
+            minNode.proofRequestIdentifier.chainId
+        ][minNode.proofRequestIdentifier.blockNumber] = 0;
+    }
+
+    function removeAt(
+        Heap storage heap,
+        IProofManager.ProofRequestIdentifier memory proofRequestIdentifier
+    ) public {
+        uint256 idx = heap.heapIndexes[
+            proofRequestIdentifier.chainId
+        ][proofRequestIdentifier.blockNumber];
+        require(idx != 0 && idx < heap.nodes.length, "Heap: idx OOB");
+
+        Node memory removed = heap.nodes[idx];
+
+        uint256 lastIdx = heap.nodes.length - 1;
+
+        if (idx == lastIdx) {
+            // Just pop last
+            heap.nodes.pop();
+            heap.heapIndexes[proofRequestIdentifier.chainId][proofRequestIdentifier.blockNumber] = 0;
+            return;
+        }
+
+        // Replace idx with last
+        Node memory moved = heap.nodes[lastIdx];
+        heap.nodes[idx] = moved;
+        heap.nodes.pop();
+
+        // Update mapping for moved node
+        heap.heapIndexes[
+            moved.proofRequestIdentifier.chainId
+        ][moved.proofRequestIdentifier.blockNumber] = idx;
+
+        // Restore heap property
+        if (idx > 1 && heap.nodes[idx].key < heap.nodes[idx / 2].key) {
+            _siftUp(heap, idx);
+        } else {
+            _siftDown(heap, idx);
```

Comment:
I believe we run `_siftDown(heap, 0)` here. Given your invariants, this looks wrong (albeit can't tell if code will break or not).